File: clients/ios/AwareChat-iOS/AwareChat-iOS/Features/Identification/IdentificationViewModel.swift
// MARK: - AI Generated - Start
import Foundation
import Observation

nonisolated struct IdentificationErrorPresentation: Equatable, Sendable {
  let message: String
  let allowsRetry: Bool
  let allowsCancel: Bool
}

nonisolated enum IdentificationState: Equatable, Sendable {
  case loading
  case form
  case error(IdentificationErrorPresentation)
}

@MainActor
@Observable
final class IdentificationViewModel {
  private(set) var state = IdentificationState.loading
  var name = "" {
    didSet { validationMessage = nil }
  }
  private(set) var validationMessage: String?

  private let users: any UserLocalRepository
  private let messaging: any MessagingServiceProtocol
  private let onRegistrationCompleted: () -> Void
  private var didLoad = false
  private var failedOperation: FailedOperation?
  private var activeAttempt: UUID?
  private var registrationTask: Task<Void, Never>?

  init(
    users: any UserLocalRepository,
    messaging: any MessagingServiceProtocol,
    onRegistrationCompleted: @escaping () -> Void
  ) {
    self.users = users
    self.messaging = messaging
    self.onRegistrationCompleted = onRegistrationCompleted
  }

  func load() async {
    guard !didLoad else { return }
    didLoad = true
    state = .loading

    do {
      let currentUser = try users.currentUser()
      name = currentUser?.name ?? ""
      if currentUser?.registrationCompleted == true {
        await messaging.start()
        onRegistrationCompleted()
      } else {
        state = .form
      }
    } catch {
      present(error, operation: .loadIdentity)
    }
  }

  func confirm() {
    guard state == .form else { return }
    guard name.contains(where: { !$0.isWhitespace }) else {
      validationMessage = "Enter a user name."
      return
    }

    beginRegistration()
  }

  func retry() {
    guard case .error(let presentation) = state, presentation.allowsRetry,
          let failedOperation else { return }

    switch failedOperation {
    case .loadIdentity:
      didLoad = false
      Task { await load() }
    case .registration:
      beginRegistration()
    }
  }

  func cancel() {
    guard case .error(let presentation) = state, presentation.allowsCancel else { return }
    invalidateAttempt()
    failedOperation = nil
    state = .form
    let messaging = messaging
    Task { await messaging.stop() }
  }

  private func beginRegistration() {
    state = .loading
    validationMessage = nil

    do {
      _ = try users.saveIdentity(name: name)
    } catch {
      present(error, operation: .registration)
      return
    }

    invalidateAttempt()
    let attempt = UUID()
    activeAttempt = attempt
    registrationTask = Task { [weak self] in
      await self?.register(attempt: attempt)
    }
  }

  private func register(attempt: UUID) async {
    let connectionStates = await messaging.observeConnectionState()
    guard isActive(attempt) else { return }
    await messaging.start()

    for await connectionState in connectionStates {
      guard isActive(attempt) else { return }

      switch connectionState {
      case .connected:
        finish(attempt)
        onRegistrationCompleted()
        return
      case .connectionFailure(let failure):
        await messaging.stop()
        guard isActive(attempt) else { return }
        finish(attempt)
        present(failure, operation: .registration)
        return
      case .disconnected, .connecting:
        continue
      }
    }
  }

  private func present(_ error: any Error, operation: FailedOperation) {
    failedOperation = operation
    state = .error(
      IdentificationErrorPresentation(
        message: error.localizedDescription,
        allowsRetry: allowsRetry(error),
        allowsCancel: operation == .registration
      )
    )
  }

  private func allowsRetry(_ error: any Error) -> Bool {
    if let failure = error as? MessagingFailure {
      switch failure {
      case .server(let serverError):
        return serverError.isRetryable
      case .storage(let persistenceError):
        return allowsRetry(persistenceError)
      case .unexpectedIdentity:
        return false
      case .network, .identificationTimedOut, .acceptanceTimedOut, .sessionReplaced:
        return true
      }
    }

    if let persistenceError = error as? PersistenceError {
      return allowsRetry(persistenceError)
    }

    return true
  }

  private func allowsRetry(_ error: PersistenceError) -> Bool {
    switch error {
    case .readFailed, .writeFailed:
      return true
    case .invalidData, .missingIdentity, .identityConflict, .messageConflict,
         .messageNotFound, .sequenceExhausted:
      return false
    }
  }

  private func isActive(_ attempt: UUID) -> Bool {
    activeAttempt == attempt && !Task.isCancelled
  }

  private func finish(_ attempt: UUID) {
    guard activeAttempt == attempt else { return }
    activeAttempt = nil
    registrationTask = nil
  }

  private func invalidateAttempt() {
    activeAttempt = nil
    registrationTask?.cancel()
    registrationTask = nil
  }

  private enum FailedOperation: Equatable {
    case loadIdentity
    case registration
  }
}
// MARK: - AI Generated - End

File: clients/android/AwareChat-Android/app/src/main/java/com/example/awarechat_android/feature/identification/IdentificationViewModel.kt
// MARK: - AI Generated - Start
package com.example.awarechat_android.feature.identification

import androidx.lifecycle.ViewModel
import androidx.lifecycle.viewModelScope
import com.example.awarechat_android.core.network.NetworkException
import com.example.awarechat_android.core.persistence.database.PersistenceException
import com.example.awarechat_android.core.persistence.database.PersistenceFailure
import com.example.awarechat_android.core.persistence.users.UserLocalRepository
import com.example.awarechat_android.core.service.MessagingConnectionState
import com.example.awarechat_android.core.service.MessagingFailure
import com.example.awarechat_android.core.service.MessagingServiceContract
import java.util.UUID
import kotlinx.coroutines.CancellationException
import kotlinx.coroutines.Job
import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.flow.StateFlow
import kotlinx.coroutines.flow.asStateFlow
import kotlinx.coroutines.flow.first
import kotlinx.coroutines.launch

data class IdentificationErrorPresentation(
    val message: String,
    val allowsRetry: Boolean,
    val allowsCancel: Boolean,
)

sealed interface IdentificationUiState {
    data object Loading : IdentificationUiState

    data class Form(
        val name: String,
        val isNameInvalid: Boolean = false,
    ) : IdentificationUiState

    data class Error(
        val name: String,
        val presentation: IdentificationErrorPresentation,
    ) : IdentificationUiState
}

class IdentificationViewModel(
    private val users: UserLocalRepository,
    private val messaging: MessagingServiceContract,
    private val onRegistrationCompleted: () -> Unit,
) : ViewModel() {
    private val mutableState = MutableStateFlow<IdentificationUiState>(IdentificationUiState.Loading)
    val state: StateFlow<IdentificationUiState> = mutableState.asStateFlow()

    private var didLoad = false
    private var failedOperation: FailedOperation? = null
    private var activeAttempt: UUID? = null
    private var operation: Job? = null

    fun load() {
        if (didLoad) return
        didLoad = true
        mutableState.value = IdentificationUiState.Loading
        operation = viewModelScope.launch {
            try {
                val current = users.currentUser()
                if (current?.registrationCompleted == true) {
                    messaging.start()
                    onRegistrationCompleted()
                } else {
                    mutableState.value = IdentificationUiState.Form(current?.name.orEmpty())
                }
            } catch (error: CancellationException) {
                throw error
            } catch (error: Throwable) {
                present(error, FailedOperation.LOAD_IDENTITY, name = "")
            }
        }
    }

    fun updateName(name: String) {
        val current = mutableState.value as? IdentificationUiState.Form ?: return
        mutableState.value = current.copy(name = name, isNameInvalid = false)
    }

    fun confirm() {
        val form = mutableState.value as? IdentificationUiState.Form ?: return
        if (form.name.isBlank()) {
            mutableState.value = form.copy(isNameInvalid = true)
            return
        }
        beginRegistration(form.name)
    }

    fun retry() {
        val error = mutableState.value as? IdentificationUiState.Error ?: return
        if (!error.presentation.allowsRetry) return

        when (failedOperation) {
            FailedOperation.LOAD_IDENTITY -> {
                didLoad = false
                load()
            }
            FailedOperation.REGISTRATION -> beginRegistration(error.name)
            null -> Unit
        }
    }

    fun cancel() {
        val error = mutableState.value as? IdentificationUiState.Error ?: return
        if (!error.presentation.allowsCancel) return
        invalidateAttempt()
        failedOperation = null
        mutableState.value = IdentificationUiState.Form(error.name)
        viewModelScope.launch { messaging.stop() }
    }

    private fun beginRegistration(name: String) {
        invalidateAttempt()
        mutableState.value = IdentificationUiState.Loading
        failedOperation = null
        val attempt = UUID.randomUUID()
        activeAttempt = attempt
        operation = viewModelScope.launch {
            try {
                users.saveIdentity(name)
                if (!isActive(attempt)) return@launch

                messaging.start()
                when (val terminal = messaging.connectionState.first {
                    it is MessagingConnectionState.Connected ||
                        it is MessagingConnectionState.ConnectionFailure
                }) {
                    MessagingConnectionState.Connected -> {
                        if (!isActive(attempt)) return@launch
                        finish(attempt)
                        onRegistrationCompleted()
                    }
                    is MessagingConnectionState.ConnectionFailure -> {
                        messaging.stop()
                        if (!isActive(attempt)) return@launch
                        finish(attempt)
                        present(terminal.failure, FailedOperation.REGISTRATION, name)
                    }
                    MessagingConnectionState.Connecting,
                    MessagingConnectionState.Disconnected,
                    -> Unit
                }
            } catch (error: CancellationException) {
                throw error
            } catch (error: Throwable) {
                if (!isActive(attempt)) return@launch
                finish(attempt)
                present(error, FailedOperation.REGISTRATION, name)
            }
        }
    }

    private fun present(error: Throwable, operation: FailedOperation, name: String) {
        showError(
            operation = operation,
            name = name,
            message = error.safeMessage(),
            allowsRetry = error.allowsRetry(),
        )
    }

    private fun present(failure: MessagingFailure, operation: FailedOperation, name: String) {
        showError(
            operation = operation,
            name = name,
            message = failure.userMessage,
            allowsRetry = failure.isRetryable,
        )
    }

    private fun showError(
        operation: FailedOperation,
        name: String,
        message: String,
        allowsRetry: Boolean,
    ) {
        failedOperation = operation
        mutableState.value = IdentificationUiState.Error(
            name = name,
            presentation = IdentificationErrorPresentation(
                message = message,
                allowsRetry = allowsRetry,
                allowsCancel = operation == FailedOperation.REGISTRATION,
            ),
        )
    }

    private fun isActive(attempt: UUID): Boolean = activeAttempt == attempt

    private fun finish(attempt: UUID) {
        if (activeAttempt != attempt) return
        activeAttempt = null
        operation = null
    }

    private fun invalidateAttempt() {
        activeAttempt = null
        operation?.cancel()
        operation = null
    }

    private enum class FailedOperation {
        LOAD_IDENTITY,
        REGISTRATION,
    }
}

private fun Throwable.safeMessage(): String = when (this) {
    is PersistenceException -> message ?: NetworkException.SAFE_FALLBACK
    is NetworkException -> message ?: NetworkException.SAFE_FALLBACK
    else -> NetworkException.SAFE_FALLBACK
}

private fun Throwable.allowsRetry(): Boolean = when (this) {
    is PersistenceException -> failure == PersistenceFailure.READ_FAILED ||
        failure == PersistenceFailure.WRITE_FAILED
    else -> true
}
// MARK: - AI Generated - End
